contents/security/crossSiteRequestForgery/crossSiteRequestForgery.md

```diff
@@ -0,0 +1,96 @@
+<input type="hidden" name="amount" value="100000"/>
+<input type="submit" value="View more pictures"/>
+</form>
+```
+
+So, when she clicks that button, the browser actually sends the following request to `http://www.example-bank.com`:
+
+```
+POST http://www.example-bank.com/transfer HTTP/1.1
+Cookie: auth=1abcd2ek3292fsa390sdf
+
+account-no=987654&amount=100000
+```
+
+Since Alice is still logged in to `www.example-bank.com`, any request made to `www.example-bank.com` is automatically attached the cookie `auth=1abcd2ek3292fsa390sdf`. Hence, the bank's server considers this as an authenticated request from Alice and proceeds to transfer $100,000 from Alice's account to account number 987654, which belongs to the attacker! Indeed, `www.i-am-not-evil.com` could have performed the attack by using Javascript to submit the form, which would eliminate the need for Alice to click on anything, so she will not even be aware of it.
+
+## Conditions for a successful CSRF attack
+
+From the example, we can see that a CSRF attack works by forging a valid request which **inherits the identity and privileges of the victim** to the vulnerable web server. Such inheritance of privileges is possible because the browser attaches any credentials associated with a website to all requests made to the site. Therefore, if the user is currently authenticated to the site, the site will have no way to distinguish between the forged request sent by the victim and a legitimate request sent by the victim. In short, there are 2 conditions necessary for the attack:
+
+1. The target website is vulnerable to a CSRF attack (i.e. it cannot distinguish between forged and legitimate requests).
+2. The victim is authenticated to the target website.
+3. The victim is tricked to visit an untrusted website by social engineering.
+
+## Defense against CSRF attacks
+
+Defense against CSRF can come from both the website and the user. As the user, one can reduce the likelihood of being attacked by logging out of sensitive services like banking immediately after carrying out transactions. This will negate condition number 2.
+
+For the web server, it can eliminate condition number 1 by adding a secret token as a hidden field to all forms that it serves. This token should be random and cannot be guessed by the attacker. Back to our example, the bank website should server a form like the following:
+
+```
+<form action="/transfer" method="POST">
+<input type="hidden" name="csrf-token" value="abcd23ksk3l2faad2kdkl"/>
+<input type="number" name="account-no" value="123456"/>
+<input type="number" name="amount" value="100"/>
+<input type="submit"/>
+</form>
+```
+
+So a valid request by Alice made from the bank's own website will also carry this secret token. Forged requests from other websites will not know the token. Thus, when the bank's server receives a request, it can distinguish between a legitimate and a forged request by checking that the CSRF token is present and valid.
```

Comment:
I didn't quite understand "So a valid request by Alice made from the bank's own website will also carry this secret token. Forged requests from other websites will not know the token" in my first reading...perhaps you might want to reiterate that this is because the server adds the hidden field. 

Maybe something like: "So a valid request made by Alice from the bank's own website will carry a secret token added by the web server, but forged requests from other websites will not as they do not have knowledge of this secret token."